### PR TITLE
[Azure] Add Monitor App Service Logs data stream

### PR DESCRIPTION
**Pull request by lucian-ioan**
## What does this PR do?

Adds App Service data stream with support for the following log categories:

- [x] HTTP logs
- [x] App Service Console Logs
- [x] App Service Application Logs
- [x] Access Audit Logs
- [x] IPSecurity Audit logs
- [x] App Service Platform logs

## Checklist

- [ ] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] I have verified that all data streams collect metrics or logs.
- [ ] I have added an entry to my package's `changelog.yml` file.
- [ ] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

Document providing instruction on how to collect each log category:
[https://docs.google.com/document/d/1AJREQvAE2mc1QU19YKgZ4QJB6OHsyw8fy_WlLRYJV74](https://docs.google.com/document/d/1AJREQvAE2mc1QU19YKgZ4QJB6OHsyw8fy_WlLRYJV74)

- [x] Research how to collect HTTP logs
- [x] Research how to collect App Service Console Logs
- [x] Research how to collect App Service Application Logs
- [x] Research how to collect Access Audit Logs
- [x] Research how to collect IPSecurity Audit logs
- [x] Research how to collect App Service Platform logs

## Related issues

-Relates elastic/obs-infraobs-team#826

### CODE REVIEW

File: packages/azure/data_stream/appservices/agent/stream/azure-eventhub.yml.hbs [hunk 1]
```diff
@@ -0,0 +1,39 @@
+{{#if connection_string}}
+connection_string: {{connection_string}}
+{{/if}}
+{{#if storage_account_container }}
+storage_account_container: {{storage_account_container}}
+{{else}}
+{{#if eventhub}}
+storage_account_container: filebeat-firewalllogs-{{eventhub}}
```
Comment (zmoog) [inline packages/azure/data_stream/appservices/agent/stream/azure-eventhub.yml.hbs:8]: I believe this `firewalllogs` is a leftover, isn't it?

File: packages/azure/data_stream/appservices/elasticsearch/ingest_pipeline/appservice-auditlogs-inner-pipeline.yml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      target_field: azure.appservices.properties.user_display_name
+  - convert:
+      field: azure.appservices.properties.user_address
+      type: ip
+      ignore_missing: true
+  - remove:
+      field:
+        - azure.appservices.Properties
+      ignore_missing: true
+on_failure:
```
Comment (zmoog) [inline packages/azure/data_stream/appservices/elasticsearch/ingest_pipeline/appservice-auditlogs-inner-pipeline.yml:34]: By removing the `azure.appservices.Properties` field, we risk dropping sub-fields that we are unaware of or Azure may add in the future.

What if we rename `Properties` into `properties` first and then rename the other fields? We won't lose unexpected field this way.

File: packages/azure/data_stream/appservices/fields/fields.yml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+          description: Substatus error code on HTTP request.
+        - name: sc_win32status
+          type: keyword
+          description: Windows status code on HTTP request.
+        - name: time_taken
+          type: long
+          description: Time taken by HTTP request in milliseconds.
+        - name: user_agent
+          type: keyword
+          description: User agent on HTTP request.
```
Comment (zmoog) [inline packages/azure/data_stream/appservices/fields/fields.yml:78]: What about grouping the fields in this list by log categories?

Adding comments to visually separate the shared ones from those belonging to a specific log category can be helpful.

**Review by zmoog — COMMENTED** [review]
This first iteration on `AppServiceAuditLogs` and `AppServiceHTTPLogs` looks promising.

I added a few comments I'd love to hear you about.

**Comment by zmoog** [comment]
What log categories are you working on right now?

**Comment by zmoog** [comment]
Let me know which log categories you are working on, and I'll try to pick at least one or two of the other ones to find the sample logs.

**Comment by lucianpy** [comment]
Currently, I have samples for the 4 log categories marked and will update this PR to contain the yaml for the other 2.

**Comment by zmoog** [comment]
Hey @lucianpy, how is the log hunting going? Please let me know if you need help with the log collection, testing, or field/pipeline building!

**Comment by lucianpy** [comment]
Hey @zmoog, no new progress as of yet. Worst case I will just push the 4 categories that I do have. Still want to try to get them this week. If not succesful, I will ask for your help on Monday.

**Comment by zmoog** [comment]
> Hey @zmoog, no new progress as of yet. Worst case I will just push the 4 categories that I do have. Still want to try to get them this week. If not succesful, I will ask for your help on Monday.

Hey @lucianpy, can I help with something?

**Comment by lucianpy** [comment]
Hey @zmoog, could you take a look at IPSecurity Audit logs? I've managed to get App Service Application Logs at some point but couldn't figure out why and couldn't get it again

**Comment by lucianpy** [comment]
As an update here, I now have all of the 6 log categories that we need. Unfortunately, `App Service Platform Logs` and `Access Audit Logs` can't be processed directly due to double newlines apearing in some fields similar to:
https://github.com/elastic/beats/issues/34092 

Currently looking at a fix for this issue.

**Comment by lucianpy** [comment]
One thing that needs to be decided is what to do in case of same name fields for different log categories.
For example, the `CIp` field for `HTTPLogs` is just a usual IP address whereas the `CIp` for `IPSecAuditLogs` is a socket (ip:port).

A solution that I found is to modify the fields for each log category, such as `azure.appservices.client_id` becomes `azure.appservices.httplogs.client_id`. The disadvantage to this is that the json that ends un in Kibana will be more verbose. What do you think @zmoog ?

Also, updated the PR with the document providing instructions on how to collect all the log categories!

**Comment by zmoog** [comment]
Quick update after the sync over zoom in the last few days to discuss the following two topics.

***

> One thing that needs to be decided is what to do in case of same name fields for different log categories.

We decided to try to map the source document into a destination field as an object. 

For example, if the source has an IP address only, we can use a destination field like this:

```json
{
  "ip": {
    "address": "10.1.0.1"
  }
}
```

Instead, if it has an IP address and a port, we can use:

```json
{
  "ip": {
    "address": "10.1.0.1",
    "port": 8080
  }
}
```

This is only an example, not the recommended names 😄 , I suggest exploring what other integration are doing and the reference names and structure in the Elastic Common Schema (ECS).

***

> Unfortunately, App Service Platform Logs and App Service Console Logs can't be processed directly due to double newlines appearing in some fields similar to: https://github.com/elastic/beats/issues/34092

We need to reproduce this issue to share our findings with the Azure folks to understand this issue better.

Suppose this turns out to be something we can't avoid. In that case, one possible approach is the update the `azure-eventhub` input at https://github.com/elastic/beats/blob/main/x-pack/filebeat/input/azureeventhub/input.go#L167-L203, and the sanitize the `[]byte` we receive from the Event Hub before trying to unmarshal it to unpack the elements in the `records` list.

**Comment by zmoog** [comment]
@lucianpy how is it going with the newlines issue? Are you and Gabriel able to reproduce the issue again?

**Comment by lucianpy** [comment]
> Suppose this turns out to be something we can't avoid. In that case, one possible approach is the update the azure-eventhub input at https://github.com/elastic/beats/blob/main/x-pack/filebeat/input/azureeventhub/input.go#L167-L203, and the sanitize the []byte we receive from the Event Hub before trying to unmarshal it to unpack the elements in the records list.

I found I can only sanitize the event for appservices, since this is where the issue was found. The advantage of this is not modifying the azure-eventhub code directly in beats. What do you think @zmoog?

**Comment by zmoog** [comment]
> I found I can only sanitize the event for appservices, since this is where the issue was found. The advantage of this is not modifying the azure-eventhub code directly in beats. What do you think @zmoog?

There is one main issue in trying to "patch" the invalid JSON string in the pipeline vs. the input.

The diagnostic setting sends data to the event hub in batches, so the input receives something like this:

```
{
  "records": [
    {
      "time": "2023-03-08T10:33:13.8810000Z",
      "resourceId": "/SUBSCRIPTIONS/6b173463-3d87-43c5-98bb-90af439f1be0/RESOURCEGROUPS/WHATEVER/PROVIDERS/MICROSOFT.INSIGHTS/COMPONENTS/WHATEVER-2",
      "ResourceGUID": "0e1fdb93-b4b6-468b-9408-70ee34e56b22",
      "Type": "AppRequests",
      "AppRoleInstance": "Maurizios-MacBook-Pro.local",
      "AppRoleName": "Web",
      "ClientIP": "0.0.0.0",
      "ClientOS": "Darwin 22.3.0",
      "ClientType": "PC",
      "IKey": "89c2be53-f9c7-4ac0-8191-fb31898b774c",
      "_BilledSize": 976,
      "OperationName": "GET /",
      "OperationId": "ba20ab37cafe455c96db689d52c6b4a9",
      "ParentId": "ba20ab37cafe455c96db689d52c6b4a9",
      "SDKVersion": "node:2.3.5",
      "Properties": {
        "_MS.ProcessedByMetricExtractors": "(Name:'Requests', Ver:'1.1')"
      },
      "Id": "|ba20ab37cafe455c96db689d52c6b4a9.",
      "Name": "GET /",
      "Url": "http://localhost:3000/",
      "Success": true,
      "ResultCode": "200",
      "DurationMs": 1,
      "PerformanceBucket": "<250ms",
      "ItemCount": 1
    },
    {
      "time": "2023-03-08T10:33:13.8810000Z",
      "resourceId": "/SUBSCRIPTIONS/6b173463-3d87-43c5-98bb-90af439f1be0/RESOURCEGROUPS/WHATEVER/PROVIDERS/MICROSOFT.INSIGHTS/COMPONENTS/WHATEVER-2",
      "ResourceGUID": "0e1fdb93-b4b6-468b-9408-70ee34e56b22",
      "Type": "AppRequests",
      "AppRoleInstance": "Maurizios-MacBook-Pro.local",
      "AppRoleName": "Web",
      "ClientIP": "0.0.0.0",
      "ClientOS": "Darwin 22.3.0",
      "ClientType": "PC",
      "IKey": "89c2be53-f9c7-4ac0-8191-fb31898b774c",
      "_BilledSize": 976,
      "OperationName": "GET /",
      "OperationId": "ba20ab37cafe455c96db689d52c6b4a9",
      "ParentId": "ba20ab37cafe455c96db689d52c6b4a9",
      "SDKVersion": "node:2.3.5",
      "Properties": {
        "_MS.ProcessedByMetricExtractors": "(Name:'Requests', Ver:'1.1')"
      },
      "Id": "|ba20ab37cafe455c96db689d52c6b4a9.",
      "Name": "GET /",
      "Url": "http://localhost:3000/",
      "Success": true,
      "ResultCode": "200",
      "DurationMs": 1,
      "PerformanceBucket": "<250ms",
      "ItemCount": 1
    }
  ]
}
```

In this case, we have two records. 

When the input fails to parse the JSON, it packs the whole `{ "records": [...] }` batch of two records into the `message` field and sends it to the data stream as one document, instead of creating one two documents (one for each record).

In a case like this, the pipeline receives one document with two records, and I am not sure if it is possible to create two documents out of it. My understaning is that the pipeline in:out ratio is 1:1.

In this example, we have two records, but it is common to have ten or more records in a batch.

**Comment by lucianpy** [comment]
@zmoog, took a second look at your example and you are right, there are issues with doing it directly from integrations.

Should I open a PR in [beats](https://github.com/elastic/beats) for this?

**Comment by zmoog** [comment]
> @zmoog, took a second look at your example and you are right, there are issues with doing it directly from integrations.
> 
> Should I open a different PR in [beats](https://github.com/elastic/beats) for this?

I believe we need to tackle this at the input level with a PR on Beats, but it’s better to double-check; let's ask other team members and see what's their advice.

**Comment by lucianpy** [comment]
As an update here, the consensus of the team is to takle this directly in Beats.

Currently waiting on the [sanitization]( https://github.com/elastic/beats/pull/34874) PR there to merge before moving forward.

**Review by zmoog — APPROVED** [review]
🚢

**Review by zmoog — COMMENTED** [review]
During a chat with @lucian-ioan, @gpop63, and @devamanv, we agreed that the best way forward for this App Service integration is the following:

-  Move the App Services integration from Azure Logs into an independent integration package.
- Like the Azure Functions integration, release it as:
  - maturity level: experimental 
  - version: v0.0.1
  - no dashboards

This approach allow us to:
- Add a metrics data stream in the App Services package so that we can have everything about App Services in a single package.
- Have the option to add a dedicated data stream for each log category instead of using a single data stream as of today.
- Collect user feedback to adjust field mapping, ingest pipelines, and dashboards.

@lalit-satapathy @rameshelastic @SubhrataK  let us know if you think we should take a different route. We plan to move forward with this in the current sprint.

**Review by hooooh2 — COMMENTED** [review]
{{#signal.rule.references}} {{.}} {{/signal.rule.references}}

**Pull request merged**